text/0000-tracked-properties.md

```diff
@@ -227,34 +248,350 @@ memoization) if a property shows up as a bottleneck during profiling.
+time we activate a tracked getter, it creates a global stack before it begins
+executing the getter's code, and as it executes, any other tracked wrappers
+that run add themselves to the stack.
+
+This example demonstrates how this method works, but is simplified to only work
+with tracked getters, and to only track keys. It uses the stage 2 decorator
+syntax.
+
+```js
+let PROPERTY_STACK;
+
+function tracked({ key, descriptor }) {
+  let originalGet = descriptor.get;
+
+  descriptor.get = function() {
+    // store the stack from the previous getter, if it exists
+    let previousStack = PROPERTY_STACK;
+
+    // create a new stack
+    PROPERTY_STACK = [];
+
+    // call the original get function to get the value
+    let value = originalGet.call(this);
+
+    // At this point, PROPERTY_STACK will contain all of the keys of any values
+    // that were marked as tracked and accessed when `originalGet` was called.
+    // We can use this knowledge to setup our change tracking, then push our
+    // own key and the current stack into the previous stack if it exists.
+    if (previousStack) {
+      previousStack.push(key, ...PROPERTY_STACK);
+    }
+
+    PROPERTY_STACK = previousStack;
+
+    return value;
+  }
+}
+```
+
+Of course, tracking keys would only work if we knew the full paths, or were only
+concerned with only changes to the current object. The real implementation uses
+[Glimmer VM's internal validators](https://github.com/glimmerjs/glimmer-vm/blob/master/guides/05-validators.md)
+to track changes to properties.
+
+As properties are accessed, their validation  _tags_ are pushed on the stack,
+and when the property access is complete these tags are combined into a single
+tag for that property. Setters are added to tracked getters and tracked simple
+properties as appropriate which manually dirty these tags whenever they are
+activated, and subsequently invalidate all of the tags for any getters which
+ever accessed them.
+
+### Untracked Properties
+
+Revisiting the `middleName` bug from earlier, you may notice that it's still
+possible for a new developer to introduce a bug with tracked properties. What if
+they don't know that all mutable properties must be tracked? This would fail,
+just like our previous example:
+
+```js
+import { tracked } from '@ember/object';
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+  middleName = 'Tomster';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.middleName} ${this.lastName}`;
+  }
+}
+```
+
+The last piece of the puzzle here will be a development time only assertion.
+When you access _untracked_ properties from an a tracked getter, we will install
+something similar to the mandatory setter function. This will throw an error if
+the user ever tries to set the property, effectively making it immutable.
+
+To accomplish this, we can wrap any object which has tracked properties with a
```

Comment:
Discussing this right now, but basically I think

1. Whenever a class is marked with `@tracked`, it gets wrapped in a proxy (we can use finishers to do this I believe)
2. Whenever a proxied object accesses and returns a property which is _also_ an object, it will wrap that object in another proxy. This means that all accesses from a tracked context will be wrapped in proxies, no matter how many layers deep.